pallets/pallet-deposits/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+
+			// emit event
+			Self::deposit_event(Event::DepositPaid(deposit_id, sender, new_total));
+
+			Ok(())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::announce_take())]
+		pub fn announce_take(
+			origin: OriginFor<T>,
+			deposit_id: DepositIdOf<T>,
+			amount: BalanceOf<T>,
+			block: <T as frame_system::Config>::BlockNumber,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			
+			// check minimal notice period
+			let current_block = frame_system::Pallet::<T>::block_number();
+			if block.saturating_sub(current_block) < T::MinimalNoticePeriod::get() {
+				return Err(Error::<T>::NoticePeriodTooShort.into());
+			}
+
+			// create the announcement
+			let announcment = DepositTakeAnnouncement {
+				block,
+				amount,
+			};
+			Announcments::<T>::insert(&deposit_id, &sender,  announcment);
+			
+			// emit event
+			Self::deposit_event(Event::DepositTakeAnnouncement(deposit_id, sender, block, amount));
+			
+			Ok(())
+		}
+		/// Take a deposit
+		/// # Arguments
+		/// * `origin`: The account that is taking the deposit
+		/// * `deposit_id` - The id of the deposit
+		#[pallet::weight(<T as Config>::WeightInfo::take())]
+		pub fn take(origin: OriginFor<T>, deposit_id: DepositIdOf<T>) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			// check if the take was announced before
+			let announcment = Announcments::<T>::get(&deposit_id, &sender).ok_or(Error::<T>::NotAnnounced)?;
+			let current_block = frame_system::Pallet::<T>::block_number();
+			if current_block < announcment.block {
+				return Err(Error::<T>::StillInNoticePeriod.into());
+			}
+
+			// check if sender has paid more or equal amount of deposit as requested to take
+			let amount = announcment.amount;
+			let sender_balance = DepositsByAccount::<T>::get(&deposit_id, &sender).unwrap_or_default();
+			ensure!(sender_balance <= amount, Error::<T>::InsufficientFunds);
+
+			// unreserve the deposit
+			<T::Currency as ReservableCurrency<AccountIdOf<T>>>::unreserve(&sender, amount);
+
+			// substract the amount from the balance for this account
+			DepositsByAccount::<T>::mutate(&deposit_id, &sender, |balance| {
+				*balance = Some(balance.unwrap_or_default().saturating_sub(amount))
+			});
+
+			// update the total amount for this deposit pot
+			let new_total = DepositsTotal::<T>::mutate(&deposit_id, |total| {
+				*total = Some(total.unwrap_or_default().saturating_sub(amount));
+				total.unwrap()
+			});
+
+			// cleanup the announcement
+			Announcments::<T>::remove(&deposit_id, &sender);
+
+			// call the change handler with the new total
+			<T::ChangeHandler>::on_deposit_taken(&deposit_id, &sender, new_total);
+
+			// emit event
+			Self::deposit_event(Event::DepositTaken(deposit_id, sender, new_total));
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// retrieve the total amount of a deposit pot
+		pub fn check_deposit(deposit_id: &DepositIdOf<T>) -> BalanceOf<T> {
```

Comment:
Where are this and `ensure_deposit` functions used? They are not called anywhere inside the pallet, and all they do is accessing the storage of this pallet, so I think we can get rid of the trait altogether.